iceoryx_posh/include/iceoryx_posh/internal/popo/client_impl.inl

```diff
@@ -0,0 +1,81 @@
+// Copyright (c) 2021 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_POSH_POPO_CLIENT_IMPL_INL
+#define IOX_POSH_POPO_CLIENT_IMPL_INL
+
+#include "iceoryx_posh/internal/popo/client_impl.hpp"
+
+namespace iox
+{
+namespace popo
+{
+template <typename Req, typename Res, typename BaseClientT>
+ClientImpl<Req, Res, BaseClientT>::ClientImpl(const capro::ServiceDescription& service,
+                                              const ClientOptions& clientOptions) noexcept
+    : BaseClientT(service, clientOptions)
+{
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+cxx::expected<Request<Req>, AllocationError> ClientImpl<Req, Res, BaseClientT>::loanUninitialized() noexcept
+{
+    auto result = port().allocateRequest(sizeof(Req), alignof(Req));
+    if (result.has_error())
+    {
+        return cxx::error<AllocationError>(result.get_error());
+    }
+    auto requestHeader = result.value();
+    auto payload = mepoo::ChunkHeader::fromUserHeader(requestHeader)->userPayload();
+    auto request = cxx::unique_ptr<Req>(reinterpret_cast<Req*>(payload), m_requestDeleter);
+    return cxx::success<Request<Req>>(Request<Req>{std::move(request), *this});
+}
+
+template <typename Req, typename Res, typename BaseClientT>
+template <typename... Args>
+cxx::expected<Request<Req>, AllocationError> ClientImpl<Req, Res, BaseClientT>::loan(Args&&... args) noexcept
+{
+    return std::move(
```

Comment:
Not sure why we need to move into the expected (return type). Ideally all should work directly with return value optimization. As far as I know these kind of move conversions are antipattern as they interfere with RVO or at least should not be required. There is also no real move happening from what  understand (expected owns its memory).

Reply:
You are right. This was again a c&p from the `Publisher`. I guess it should be changed there too

Reply:
Ok, then I think we will leave it as is. But I o not understand why `SmartChunk` is movable but not copyable. I would imagine it could have multiple (read) owners in a proper design.

Reply:
@MatthiasKillat I'm not sure if it was done in the initial implementation of the `Sample` or later on. When a `Publisher` creates a `Sample` it might make sense to not copy it but for a `Subscriber` it most probably should be possible. For this, we would need to replace the `cxx::unique_ptr` with a `cxx::shared_ptr`, which will be interesting to be implemented without the heap.

Reply:
It is possible to implement, even without dynamic memory (if I did not miss something, only done mentally), but not pretty or efficient without a factory which can indirectly store the refcounts.